src/box/tx_memory.c

```diff
@@ -0,0 +1,195 @@
+#include <assert.h>
+#include <lib/core/say.h>
+#include "tx_memory.h"
+#include "small/mempool.h"
+#include "small/region.h"
+#include "histogram.h"
+#include "txn.h"
+
+const char *TXN_ALLOC_TYPE_STRS[] = {
+	"STATEMENTS",
+	"SAVEPOINTS",
+	"USER DATA",
+	"REDO LOGS",
+	"TRIGGERS",
+	"TIMERS",
+	"JOURNAL ENTRIES"
+};
+
+static_assert(lengthof(TXN_ALLOC_TYPE_STRS) == TXN_ALLOC_MAX,
+	      "TXN_ALLOC_TYPE_STRS does not match TXN_ALLOC_TYPE");
+
+static void
+tx_track_allocation(struct tx_memory_manager *stat, struct txn *txn,
+		    uint64_t alloc_size, size_t alloc_type, bool deallocate)
+{
+	assert(alloc_type < stat->alloc_max);
+	assert(txn != NULL);
+
+	histogram_discard(stat->stats_storage[alloc_type].hist,
+			  txn->mem_used->total[alloc_type]);
+	if (!deallocate) {
+		txn->mem_used->total[alloc_type] += alloc_size;
+		stat->stats_storage[alloc_type].total += alloc_size;
+	} else {
+		assert(txn->mem_used->total[alloc_type] >= alloc_size);
+		assert(stat->stats_storage[alloc_type].total >= alloc_size);
+		txn->mem_used->total[alloc_type] -= alloc_size;
+		stat->stats_storage[alloc_type].total -= alloc_size;
+	}
+	histogram_collect(stat->stats_storage[alloc_type].hist,
+			  txn->mem_used->total[alloc_type]);
+}
+
+void
+tx_memory_register_txn(struct tx_memory_manager *stat, struct txn *txn)
+{
+	assert(txn != NULL);
+	assert(txn->mem_used == NULL);
+	assert(txn->given_region_used == 0);
+
+	struct txn_mem_used *mem_used =
+		region_aligned_alloc(&txn->region,
+				     sizeof(struct txn_mem_used)  +
```

Comment:
1. You can move out size of allocation calculation to a separate var to save a few lines.
2. You don't check the result of region alloc (still it can fail)

Reply:
1. Done
2. Fixed